numpy/random/mtrand.pyx

```diff
@@ -779,7 +779,9 @@ cdef class RandomState:
                 if np.issubdtype(p.dtype, np.floating):
                     atol = max(atol, np.sqrt(np.finfo(p.dtype).eps))
 
-            p = <np.ndarray>np.PyArray_FROM_OTF(p, np.NPY_DOUBLE, np.NPY_ALIGNED)
+            p = <np.ndarray>np.PyArray_FROM_OTF(
+                p, np.NPY_DOUBLE, np.NPY_ALIGNED | np.NPY_ARRAY_C_CONTIGUOUS
+            )
```

Comment:
would be better to change the `kahan_sum` function rather than forcing a copy here. There are more places that need testing/fixing

Reply:
something like passing a memoryview instead of a pointer to `kahan_sum` ?

Reply:
I think it has to go here (and in `Generator`) since p has to be converted from other iteable containers anyway, The downside of having two very similar files.

If this also appears in multinomial, then it would be best to create a new function in `common.pyx`, that would do the conversion and the checks (or at least as many as can be done).